Add new -timeout command-line option

aryx opened the pull request:
This closes issue #811

Test plan:
time yy -timeout 2 -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
+ /home/pad/github/semgrep/semgrep-core/_build/default/bin/Main.exe -equivalences /home/pad/github/semgrep/semgrep-core/data/basic_equivalences.yml -timeout 2 -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
{
  "matches": [],
  "errors": [
    {
      "check_id": "FatalError",
      "path": "tests/PERF/timeout.js",
      "start": { "line": 1, "col": 1 },
      "end": { "line": 1, "col": 1 },
      "extra": {
        "message": "Fatal Error: Timeout",
        "line": "ACC.captcha = {"
      }
    }
  ],
  "stats": { "okfiles": 0, "errorfiles": 1 }
}

real	0m2.078s
user	0m2.018s
sys	0m0.053s

aryx committed:
Add new -timeout command-line option

This closes issue #811

Test plan:
time yy -timeout 2 -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
+ /home/pad/github/semgrep/semgrep-core/_build/default/bin/Main.exe -equivalences /home/pad/github/semgrep/semgrep-core/data/basic_equivalences.yml -timeout 2 -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
{
  "matches": [],
  "errors": [
    {
      "check_id": "FatalError",
      "path": "tests/PERF/timeout.js",
      "start": { "line": 1, "col": 1 },
      "end": { "line": 1, "col": 1 },
      "extra": {
        "message": "Fatal Error: Timeout",
        "line": "ACC.captcha = {"
      }
    }
  ],
  "stats": { "okfiles": 0, "errorfiles": 1 }
}

real	0m2.078s
user	0m2.018s
sys	0m0.053s

ievans commented:
Can check_id be "TimeoutError" rather than FatalError? Would be easier to dispatch on any UI stuff in the frontend.

ievans commented:
Are timeouts supported on all advertised "supported" parsers @pad?

aryx committed:
Put back timeout for Javascript Parsing

The semgrep CI currently take more than 1h because we removed the
5s timeout for Javascript parsing in pfff.
This diff puts it back in semgrep, to get the same old behavior.

Test plan:
time yy -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
+ /home/pad/github/semgrep/semgrep-core/_build/default/bin/Main.exe -equivalences /home/pad/github/semgrep/semgrep-core/data/basic_equivalences.yml -lang js -rules_file data/basic.yml -json tests/PERF/timeout.js
{
  "matches": [],
  "errors": [
    {
      "check_id": "Timeout",
      "path": "tests/PERF/timeout.js",
      "start": { "line": 1, "col": 1 },
      "end": { "line": 1, "col": 1 },
      "extra": { "message": "Timeout", "line": "ACC.captcha = {" }
    }
  ],
  "stats": { "okfiles": 0, "errorfiles": 1 }
}

real	0m5.106s
user	0m5.029s
sys	0m0.070s
